src/ddscxx/tests/WaitSet.cpp

```diff
@@ -928,4 +947,131 @@ TEST_F(WaitSet, status_condition_trigger)
 
   EXPECT_EQ(dr.requested_incompatible_qos_status().total_count(), 1);
   EXPECT_EQ(dw.offered_incompatible_qos_status().total_count(), 1);
-}
\ No newline at end of file
+}
+
+/**
+ * Check that it is possible to close the conditions while waiting and then waitset
+ * automatically detaches the conditions from the waitset
+ */
+TEST_F(WaitSet, close_during_wait)
+{
+  // close read condition
+  {
+    waitSet = dds::core::cond::WaitSet();
+    auto readCondition = dds::sub::cond::ReadCondition(reader,
+                                                       dds::sub::status::DataState::any_data());
+    EXPECT_NO_THROW(waitSet.attach_condition(readCondition));
+
+    // close the read condition while waiting
+    actionThreadArgs.action = TA_CLOSE_READ_CONDITION;
+    actionThreadArgs.readCondition = &readCondition;
+    ddsrt_thread_create(&threadId, "test_action_thread",
+                        &threadAttr, test_action_thread, &actionThreadArgs);
+
+    // since there is no data, waiting for data should result in timeout
+    EXPECT_THROW(waitSet.wait(dds::core::Duration(2,0)), dds::core::TimeoutError);
+
+    // closing the reader should detach the conditions from the waitset
+    auto conditionList = waitSet.conditions();
+    EXPECT_EQ(conditionList.size(), 0);
+    ddsrt_thread_join(threadId, nullptr);
+  }
+
+  // can't really close a status condition, enable the test after the fix
+// close status condition
+//  {
+//    waitSet = dds::core::cond::WaitSet();
+//    auto statusCondition = dds::core::cond::StatusCondition(reader);
+//    statusCondition.enabled_statuses(dds::core::status::StatusMask::none());
+//    waitSet.attach_condition(statusCondition);
+//
+//    // close the read condition while waiting
+//    actionThreadArgs.action = TA_CLOSE_STATUS_CONDITION;
```

Comment:
Closing a status condition will make it unusable, so commented this block until its fixed

Reply:
As closing a StatusCondition is dubious in the first place, I'm fine with this approach.